### PR TITLE
Allow using '@' in object (key) names

### PR DESCRIPTION
**Pull request by dkasyanov**
Related to the [#94](https://github.com/RaRe-Technologies/smart_open/issues/94) issue.

Keys (objects) in AWS S2 may contain multiple '@'. 
So it may be more than 2 '@' in the URI.

P.S. Bucket names can contain only lowercase letters, numbers, and hyphens.

### CODE REVIEW

**Comment by dkasyanov** [comment]
Sure, I'll add them today.

**Comment by dkasyanov** [comment]
@mpenkov I've added tests. 
Checked that bucket name cannot contain '@'. Such test was added also.

File: smart_open/tests/test_smart_open.py [hunk 1]
```diff
@@ -83,7 +83,28 @@ def test_s3_uri(self):
         self.assertEqual(parsed_uri.access_id, "accessid")
         self.assertEqual(parsed_uri.access_secret, "access/secret")
 
-        # incorrect uri - only two '@' in uri are allowed
+        # correct uri, contains credentials and '@' in object name
```
Comment (mpenkov) [inline smart_open/tests/test_smart_open.py]: This should be a separate test.

File: smart_open/tests/test_smart_open.py [hunk 2]
```diff
@@ -83,7 +83,28 @@ def test_s3_uri(self):
...
+        parsed_uri = smart_open_lib._parse_uri("s3://accessid:access/secret@mybucket/my@ke@y")
+        self.assertEqual(parsed_uri.scheme, "s3")
+        self.assertEqual(parsed_uri.bucket_id, "mybucket")
+        self.assertEqual(parsed_uri.key_id, "my@ke@y")
+        self.assertEqual(parsed_uri.access_id, "accessid")
+        self.assertEqual(parsed_uri.access_secret, "access/secret")
+
+        # correct uri, contains credentials, host, port and '@' in object name
+        parsed_uri = smart_open_lib._parse_uri("s3://accessid:access/secret@hostname:1234@mybucket/dir/my@ke@y")
```
Comment (mpenkov) [inline smart_open/tests/test_smart_open.py]: This should be a separate test.

File: smart_open/tests/test_smart_open.py [hunk 3]
```diff
@@ -83,7 +83,28 @@ def test_s3_uri(self):
...
+        self.assertEqual(parsed_uri.scheme, "s3")
+        self.assertEqual(parsed_uri.bucket_id, "mybucket")
+        self.assertEqual(parsed_uri.key_id, "dir/my@ke@y")
+        self.assertEqual(parsed_uri.access_id, "accessid")
+        self.assertEqual(parsed_uri.access_secret, "access/secret")
+        self.assertEqual(parsed_uri.host, "hostname")
+        self.assertEqual(parsed_uri.port, 1234)
+
+        # incorrect uri - bucket can't contain '@'
```
Comment (mpenkov) [inline smart_open/tests/test_smart_open.py]: This should be a separate test.

**Comment by menshikh-iv** [comment]
ping @dkasyanov, are you plan to finish PR?

**Comment by mpenkov** [comment]
@menshikh-iv We can probably merge this and clean up the tests as a separate ticket.

My main issue was that each of the new unit tests tests multiple things, which is bad unit testing practice.

Alternatively, I can finish this PR myself. Let me know which you prefer.

**Comment by menshikh-iv** [comment]
@mpenkov it will be nice if you finish PR yourself

**Comment by menshikh-iv** [comment]
Continued in #224

**Pull request closed without merging**